native/explorer/src/encoding.rs

```diff
@@ -371,54 +370,32 @@ fn time_series_to_list<'b>(s: &Series, env: Env<'b>) -> Result<Term<'b>, Explore
     ))
 }
 
-fn generic_binary_series_to_list<'a, 'b, T, G>(
+fn generic_string_series_to_list<'b>(s: &Series, env: Env<'b>) -> Result<Term<'b>, ExplorerError> {
+    Ok(unsafe_iterator_series_to_list!(
+        env,
+        s.str()?.into_iter().map(|option| option.encode(env))
+    ))
+}
+
```

Comment:
i am not confident of `generic_string_series_to_list` and `generic_binary_series_to_list`. Feel free to make it better. I have a feeling that I messed up something here.

Reply:
@josevalim this code ok ?

Reply:
I have no idea, we will need to wait for @philss' review. :)

Reply:
I am not sure about one thing performance - earlier binaries and strings were created from buffers using binary and sub_binary. Now because of the new arrow implementation, they are exposing BinaryArrayView which gives values. So not sure if there is way to get earlier implementation working.

Reply:
I think code-wise it looks good! But I think we need a benchmark to be sure about performance.

Reply:
I am traveling this week. Earliest I can look into this next week. If anyone wants to work on this, please go ahead.

Reply:
No problem! I will take care of it. Thanks, and safe travels!

Reply:
Hey! Sorry for the delay. I could find something interesting about this.

For the test, I'm trying to compare how much time it takes to encode a binary series, and the MB of RAM it uses. I'm using a custom allocator called [`PeakAlloc`](https://github.com/xgillard/peak_alloc) for this.

The patch of the diffs I made to measure this is located in https://gist.github.com/philss/3f519ce2587461aa61d7fd9f77e3ea1f (only the Cargo.lock is out to avoid conflicts).

### Timing and memory usage - main branch

Running on `main` without calling GC:

```
$ MIX_ENV=prod mix run -e 'for i <- [:medium, :medium, :big, :medium, :medium], do: :timer.tc(fn -> Explorer.DataFrame.from_parquet!("./tmp/#{i}.parquet")["bins"] |> Explorer.Series.to_list() end) |> tap(fn {time, _val} -> IO.puts("done in: #{time / 1_000_000}") end)'
```

Results:

```
Begin: 30.768875 MB of RAM.
End: 30.769852 MB of RAM.
done in: 0.099455

Begin: 61.288383 MB of RAM.
End: 61.288383 MB of RAM.
done in: 0.101318

Begin: 3082.6296 MB of RAM.
End: 3082.6296 MB of RAM.
done in: 7.900625

Begin: 3082.6296 MB of RAM.
End: 3082.6296 MB of RAM.
done in: 2.738504

Begin: 61.288383 MB of RAM.
End: 61.288383 MB of RAM.
done in: 0.083659
```

We can see above that the memory grows, but soon it drops again when we try with a smaller file. It looks like the GC takes a while to run and drop the encoded content.

### Timing and memory usage - this PR branch

```
$ MIX_ENV=prod mix run -e 'for i <- [:medium, :medium, :big, :medium, :medium, :small, :small], do: :timer.tc(fn -> Explorer.DataFrame.from_parquet!("./tmp/#{i}.parquet")["bins"] |> Explorer.Series.to_list() end) |> tap(fn {time, _val} -> IO.puts("done in: #{time / 1_000_000}"); :erlang.garbage_collect(); Process.sleep(1_000) end)'
```

Even with GC runs (I tried before without it), it does not drop after a while:

```
Begin: 39.60874 MB of RAM.
End: 39.60972 MB of RAM.
done in: 0.106566

Begin:  78.96812 MB of RAM.
End:  78.96812 MB of RAM.
done in: 0.130126

Begin:  4662.26 MB of RAM.
End:  4662.26 MB of RAM.
done in: 9.069385

Begin:  4701.618 MB of RAM.
End:  4701.618 MB of RAM.
done in: 0.042475

Begin:  4740.9766 MB of RAM.
End:  4740.9766 MB of RAM.
done in: 0.049214

Begin:  4741.016 MB of RAM.
End:  4741.016 MB of RAM.
done in: 0.323967

Begin:  4741.0557 MB of RAM.
End:  4741.0557 MB of RAM.
done in: 0.327083
```

And if we try to load the DF without encoding the series, it is dropped after a while. See below: 

```
MIX_ENV=prod mix run -e 'for i <- [:medium, :medium, :big, :medium, :medium, :small, :small], do: :timer.tc(fn -> Explorer.DataFrame.from_parquet!("./tmp/#{i}.parquet")["bins"]; Explorer.Series.from_list([1, 2]) |> Explorer.Series.to_list() end) |> tap(fn {time, _val} -> IO.puts("done in: #{time / 1_000_000}"); Process.sleep(1_000) end)'
```

```
Begin: 39.609077 MB of RAM.
End: 39.610054 MB of RAM.
done in: 0.026808

Begin: 78.96885 MB of RAM.
End: 78.96885 MB of RAM.
done in: 0.02043

Begin: 4662.2607 MB of RAM.
End: 4662.2607 MB of RAM.
done in: 0.453696

Begin: 4701.6196 MB of RAM.
End: 4701.6196 MB of RAM.
done in: 0.012698

Begin: 4740.9785 MB of RAM.
End: 4740.9785 MB of RAM.
done in: 0.015301

Begin: 4741.0186 MB of RAM.
End: 4741.0186 MB of RAM.
done in: 3.74e-4

Begin: 0.2911501 MB of RAM.
End: 0.2911501 MB of RAM.
done in: 0.27713
```

So **I suspect the new `to_list/1` implementation for binary series is leaking something**.
My idea is to investigate more the new APIs and see if the problem is in the Polars side or ours.

PS: The parquet files used in this experiment were made by creating a DF with a column called `bins` that contains random bytes (using `:crypto.strong_rand_bytes(24)`). The size of the files is the following: 

```
-rw-r--r--. 1 philip philip 2.7G Feb 22 16:07 tmp/big.parquet (100_000_000 rows)
-rw-r--r--. 1 philip philip  27M Feb 22 15:49 tmp/medium.parquet (100_000 rows)
-rw-r--r--. 1 philip philip  28K Feb 22 15:48 tmp/small.parquet (1_000 rows)
```

Reply:
OK, I'm pretty sure that this is the expected behavior, since Polars changed its internal string/binary series representation. According to the blog post "[Why we have rewritten the string data type](https://pola.rs/posts/polars-string-type/)", they may store binaries in memory for more time (even requiring GC), and they may require more space as well. So what we had previously is not possible to achieve in this new model, for what I could understand.

Reply:
So that resolves last pending item in the review ?